ambry-clustermap/src/main/java/com.github.ambry.clustermap/DynamicClusterChangeHandler.java

```diff
@@ -0,0 +1,487 @@
+  public List<AmbryReplica> getReplicaIds(AmbryDataNode ambryDataNode) {
+    return new ArrayList<>(ambryDataNodeToAmbryReplicas.get(ambryDataNode).values());
+  }
+
+  @Override
+  public List<AmbryDataNode> getAllDataNodes() {
+    return new ArrayList<>(instanceNameToAmbryDataNode.values());
+  }
+
+  @Override
+  public Set<AmbryDisk> getDisks(AmbryDataNode ambryDataNode) {
+    return ambryDataNodeToAmbryDisks.get(ambryDataNode);
+  }
+
+  @Override
+  public Map<String, String> getPartitionToResourceMap() {
+    return Collections.unmodifiableMap(partitionNameToResource);
+  }
+
+  @Override
+  public long getErrorCount() {
+    return errorCount.get();
+  }
+
+  @Override
+  public void waitForInitNotification() throws InterruptedException {
+    // wait slightly more than 5 mins to ensure routerUpdater refreshes the snapshot.
+    if (!routingTableInitLatch.await(320, TimeUnit.SECONDS)) {
+      throw new IllegalStateException("Initial routing table change from " + dcName + " didn't come within 5 mins");
+    }
+  }
+
+  /**
+   * Add new instance or update existing instance based on {@link InstanceConfig}(s).
+   * @param instanceConfigs the {@link InstanceConfig}(s) used to update in-mem cluster map.
+   * @throws Exception
+   */
+  private void addOrUpdateInstanceInfos(List<InstanceConfig> instanceConfigs) throws Exception {
+    for (InstanceConfig instanceConfig : instanceConfigs) {
+      int schemaVersion = getSchemaVersion(instanceConfig);
+      if (schemaVersion != 0) {
+        logger.error("Unknown InstanceConfig schema version: {}, ignoring.", schemaVersion);
+        continue;
+      }
+      if (instanceNameToAmbryDataNode.containsKey(instanceConfig.getInstanceName())) {
+        updateInstanceInfo(instanceConfig);
+      } else {
+        createNewInstance(instanceConfig);
+      }
+    }
+  }
+
+  /**
+   * Update info of an existing instance. This may happen in following cases: (1) new replica is added; (2) old replica
+   * is removed; (3) replica's state has changed (i.e. becomes seal/unseal).
+   * @param instanceConfig the {@link InstanceConfig} used to update info of instance.
+   */
+  private void updateInstanceInfo(InstanceConfig instanceConfig) throws Exception {
+    String instanceName = instanceConfig.getInstanceName();
+    logger.info("Updating replicas info for existing node {}", instanceName);
+    List<String> sealedReplicas = getSealedReplicas(instanceConfig);
+    List<String> stoppedReplicas = getStoppedReplicas(instanceConfig);
+    ConcurrentHashMap<String, AmbryReplica> currentReplicasOnNode =
+        ambryDataNodeToAmbryReplicas.get(instanceNameToAmbryDataNode.get(instanceName));
+    ConcurrentHashMap<String, AmbryReplica> replicasFromConfig = new ConcurrentHashMap<>();
+    Map<String, Map<String, String>> diskInfos = instanceConfig.getRecord().getMapFields();
+    for (Map.Entry<String, Map<String, String>> entry : diskInfos.entrySet()) {
+      String mountPath = entry.getKey();
```

Comment:
I suggest check if the mountPath can map to a AmbryDisk in the ambryDataNodeToAmbryDisks first. If not, just throw an exception. We can't want to deal with the case of adding a disk.

And also check if it removes a disk or not.

Reply:
Removing disk is not in the plan but adding disk is actually supported in StorageManager. For now, we temporarily skip adding new disk in dynamic handler. We can support this if needed in the future.
I will take your suggestion to check if the mount path is valid.